### PR TITLE
Store Objects For QFunction Reassembly

### PR DESCRIPTION
**Pull request by jeremylt**
Work to reduce unneeded work for diagonal assembly, possibly with some kernel caching

### CODE REVIEW

**Comment by jeremylt** [comment]
ToDo:
```
/home/jeremy/Dev/libCEED/backends/opt/ceed-opt-operator.c:796:33: warning: passing argument 5 of ‘CeedSetBackendFunction’ from incompatible pointer type [-Wincompatible-pointer-types]
  796 |                                 CeedOperatorLinearAssembleQFunction_Opt);
      |                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                 |
      |                                 int (*)(struct CeedOperator_private *, _Bool,  struct CeedVector_private **, struct CeedElemRestriction_private **, struct CeedRequest_private **)
In file included from /home/jeremy/Dev/libCEED/backends/opt/ceed-opt-operator.c:18:
./include/ceed/backend.h:122:69: note: expected ‘int (*)()’ but argument is of type ‘int (*)(struct CeedOperator_private *, _Bool,  struct CeedVector_private **, struct CeedElemRestriction_private **, struct CeedRequest_private **)’
  122 |                                        const char *func_name, int (*f)());
      |                                                               ~~~~~~^~~~
```
I'm getting this strange garbage warning with the last commit.

File: interface/ceed-preconditioning.c [hunk 1]
```diff
@@ -1073,6 +1074,47 @@ int CeedOperatorLinearAssembleQFunction(CeedOperator op, CeedVector *assembled,
...
+int CeedOperatorLinearAssembleQFunctionUpdate(CeedOperator op,
+    CeedVector assembled, CeedElemRestriction rstr, CeedRequest *request) {
+  int ierr;
+  ierr = CeedOperatorCheckReady(op); CeedChk(ierr);
+
+  // Backend version
+  if (op->LinearAssembleQFunction) {
+    ierr = op->LinearAssembleQFunction(op, true, &assembled, &rstr, request);
+    CeedChk(ierr);
+    return CEED_ERROR_SUCCESS;
```
Comment (jedbrown) [inline interface/ceed-preconditioning.c:1105]: Am I missing something big or do you want this to be `false` so it assembles with the caller's provided vector and restriction?
```suggestion
    ierr = op->LinearAssembleQFunction(op, false, &assembled, &rstr, request);
    CeedChk(ierr);
    return CEED_ERROR_SUCCESS;
```

Comment (jeremylt) [reply interface/ceed-preconditioning.c:1105]: Yeah, that will be switched over to false. I'll be expanding this commit to hold the vector and elem restriction for operator diagonal assembly and I'll add this update at that time.

Comment (jeremylt) [reply interface/ceed-preconditioning.c]: Updated with a slightly different approach. This should help with #805 somewhat

**Comment by jeremylt** [comment]
Same warning with Clang too. I think it has to do with a bad interaction between compiler passes to handle `bool` and `int (*)()`?
```
/home/jeremy/Dev/libCEED/backends/opt/ceed-opt-operator.c:796:33: warning: incompatible function pointer types passing 'int (CeedOperator, bool, CeedVector *, CeedElemRestriction *, CeedRequest *)' (aka 'int (struct CeedOperator_private *, bool, struct CeedVector_private **, struct CeedElemRestriction_private **, struct CeedRequest_private **)') to parameter of type 'int (*)()' [-Wincompatible-function-pointer-types]
                                CeedOperatorLinearAssembleQFunction_Opt);
                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./include/ceed/backend.h:122:69: note: passing argument to parameter 'f' here
                                       const char *func_name, int (*f)());
```

The problem goes away if I use `int` in place of `bool`. Will this be the dumb thing that gets us to add a `CeedBool`?

**Comment by nbeams** [comment]
Does `bool` get promoted to `int` in the default promotions?  It looks like the same issue I ran in to for float vs double in the prototype (Issue # 1 in #784).  In that case, an explicit cast of the function pointer fixed the warnings, and due to libCEED's structure, the function pointer should never actually be called in an undefined way.

**Comment by jeremylt** [comment]
I think you might be right. I just don't really like obligating backend developers to do this silly cast.

**Comment by jeremylt** [comment]
I think it makes sense to address #805 in two parts - this simple fix and a follow-up caching kernels fix

**Comment by jedbrown** [comment]
Indeed @nbeams, that's exactly what's happening. Also notable is that `float` is promoted to `double` when calling a function with unspecified arguments (that's why `printf` can use the same format specifiers for `float` and `double`), which may be relevant to your mixed precision work.

I'm not a fan of `bool` arguments simply because they're much less self-documenting than an enum like `CEED_REUSE_CACHE` vs `CEED_BUILD_CACHE`. It's also okay to use a different function pointer and have the implementation call a `static` helper with a `bool`. That avoids the design choice to coalesce those two cases (with somehow different mutability/ownership implications) from leaking into the backend API.

**Comment by jeremylt** [comment]
I like the static helper with a bool approach. I'll revisit in the morning.

**Comment by jeremylt** [comment]
Updated

**Review by jedbrown — COMMENTED** [review]
Is this intended to remove the recompilation on each SNES iteration or is this just a prerequisite (#805)?
```diff
diff --git i/backends/cuda/ceed-cuda.c w/backends/cuda/ceed-cuda.c
index ae508bd0..ca270ddc 100644
--- i/backends/cuda/ceed-cuda.c
+++ w/backends/cuda/ceed-cuda.c
@@ -83,6 +83,7 @@ int CeedCompileCuda(Ceed ceed, const char *source, CUmodule *module,
     return CeedError(ceed, CEED_ERROR_BACKEND, "%s\n%s",
                      nvrtcGetErrorString(result), log);
   }
+  printf("Compiled program\n");

   size_t ptxsize;
   CeedChk_Nvrtc(ceed, nvrtcGetPTXSize(prog, &ptxsize));
```
then
```console
$ build/solids-elasticity -E 10 -nu .3 -degree 3 -mesh ~/meshes/holes.msh -problem FSCurrent-NH2 -num_steps 3 -bc_clamp 1 -bc_traction 2 -bc_traction_2 0,1,0 -snes_monitor -ceed /gpu/cuda/gen
[...]
  1 SNES Function norm 3.001944084211e-02
Compiled program
Compiled program
  2 SNES Function norm 1.149946571313e-03
Compiled program
Compiled program
  3 SNES Function norm 4.848972180168e-05
Compiled program
Compiled program
  4 SNES Function norm 4.482865095512e-08
Compiled program
Compiled program
  5 SNES Function norm 7.446388914027e-14
1 Load Increment
  0 SNES Function norm 2.999573779527e-02
Compiled program
Compiled program
  1 SNES Function norm 2.872648712806e-02
```

**Comment by jeremylt** [comment]
It is intended to prevent the compilation that comes from creating new objects inside of QFunction assembly. It looks like there are additional compilations to eradicate.

**Comment by jeremylt** [comment]
One of those is a work vector that is getting created and destroyed in the operator diagonal assembly on each SNES iteration.

**Comment by jeremylt** [comment]
Two thoughts:

1) How concerned should we be that we're caching rather large vectors?

2) Would the reallocation of these vectors make a significant speed difference on the CPU side?

CI failed due to whatever that intermittent OCCA issue is.

**Comment by jedbrown** [comment]
I consider one vector to be benign, especially in that it may not affect peak memory usage.

`malloc` reuses process memory (extending as needed using `sbrk`), but large allocations are done using `mmap` (a system call, and returned to the kernel when freed). The kernel generally needs to zero pages before giving them to a process (usually lazily as they're faulted). Depending on the page size, that can involve an extra pass over the data. Not a huge deal, but it's better to keep the memory.

If someone is working in a tightly memory-constrained setting, we can make an interface that releases any memory that can be rebuilt.

**Comment by jedbrown** [comment]
Can you try running
```console
$ build/solids-elasticity -E 10 -nu .3 -degree 3 -mesh ~/meshes/holes.msh -problem FSCurrent-NH2 -num_steps 3 -bc_clamp 1 -bc_traction 2 -bc_traction_2 0,1,0 -snes_monitor -ceed /gpu/cuda/shared
[...]
Compiled program
/home/jed/src/libceed/backends/cuda/ceed-cuda-operator.c:624 in CeedOperatorLinearAssembleQFunctionCore_Cuda(): Cannot assemble QFunction without active inputs and outputs
[kichatna:13991] *** Process received signal ***
[kichatna:13991] Signal: Aborted (6)
[kichatna:13991] Signal code:  (-6)
[kichatna:13991] [ 0] /usr/lib64/libpthread.so.0(+0x13870)[0x7f75b63d6870]
[kichatna:13991] [ 1] /usr/lib64/libc.so.6(gsignal+0x142)[0x7f75b6231d22]
[kichatna:13991] [ 2] /usr/lib64/libc.so.6(abort+0x116)[0x7f75b621b862]
[kichatna:13991] [ 3] /home/jed/src/libceed/lib/libceed.so(CeedErrorExit+0x0)[0x7f75b6666d53]
[kichatna:13991] [ 4] /home/jed/src/libceed/lib/libceed.so(CeedErrorImpl+0xb6)[0x7f75b66674d7]
[kichatna:13991] [ 5] /home/jed/src/libceed/lib/libceed.so(+0x3c65b)[0x7f75b668d65b]
[kichatna:13991] [ 6] /home/jed/src/libceed/lib/libceed.so(+0x3c9d6)[0x7f75b668d9d6]
[kichatna:13991] [ 7] /home/jed/src/libceed/lib/libceed.so(CeedOperatorLinearAssembleQFunctionBuildOrUpdate+0x6f)[0x7f75b666f2df]
[kichatna:13991] [ 8] /home/jed/src/libceed/lib/libceed.so(+0x3a825)[0x7f75b668b825]
[kichatna:13991] [ 9] /home/jed/src/libceed/lib/libceed.so(+0x3c043)[0x7f75b668d043]
[kichatna:13991] [10] /home/jed/src/libceed/lib/libceed.so(CeedOperatorLinearAssembleDiagonal+0x80)[0x7f75b6671222]
[kichatna:13991] [11] build/solids-elasticity(+0x20ada)[0x560c994cbada]
[kichatna:13991] [12] /home/jed/petsc/ompi-optg/lib/libpetsc.so.3.015(+0x70cef6)[0x7f75b7238ef6]
[kichatna:13991] [13] /home/jed/petsc/ompi-optg/lib/libpetsc.so.3.015(MatGetDiagonal+0x1e)[0x7f75b731d5de]
[kichatna:13991] [14] /home/jed/petsc/ompi-optg/lib/libpetsc.so.3.015(+0xbf55a5)[0x7f75b77215a5]
[kichatna:13991] [15] /home/jed/petsc/ompi-optg/lib/libpetsc.so.3.015(+0xbf5ca1)[0x7f75b7721ca1]
[kichatna:13991] [16] /home/jed/petsc/ompi-optg/lib/libpetsc.so.3.015(PCApply+0x1eb)[0x7f75b7892ccb]
[kichatna:13991] [17] /home/jed/petsc/ompi-optg/lib/libpetsc.so.3.015(+0xde4284)[0x7f75b7910284]
[kichatna:13991] [18] /home/jed/petsc/ompi-optg/lib/libpetsc.so.3.015(+0xde4788)[0x7f75b7910788]
[kichatna:13991] [19] /home/jed/petsc/ompi-optg/lib/libpetsc.so.3.015(+0xe31ce3)[0x7f75b795dce3]
[kichatna:13991] [20] /home/jed/petsc/ompi-optg/lib/libpetsc.so.3.015(KSPSolve+0x13)[0x7f75b79607b3]
[kichatna:13991] [21] /home/jed/petsc/ompi-optg/lib/libpetsc.so.3.015(+0xdd0942)[0x7f75b78fc942]
[kichatna:13991] [22] /home/jed/petsc/ompi-optg/lib/libpetsc.so.3.015(KSPSetUp+0x739)[0x7f75b795b4d9]
[kichatna:13991] [23] /home/jed/petsc/ompi-optg/lib/libpetsc.so.3.015(+0xbdbee9)[0x7f75b7707ee9]
[kichatna:13991] [24] /home/jed/petsc/ompi-optg/lib/libpetsc.so.3.015(PCSetUp+0x321)[0x7f75b78925d1]
[kichatna:13991] [25] /home/jed/petsc/ompi-optg/lib/libpetsc.so.3.015(KSPSetUp+0x3f4)[0x7f75b795b194]
[kichatna:13991] [26] /home/jed/petsc/ompi-optg/lib/libpetsc.so.3.015(+0xe31908)[0x7f75b795d908]
[kichatna:13991] [27] /home/jed/petsc/ompi-optg/lib/libpetsc.so.3.015(KSPSolve+0x13)[0x7f75b79607b3]
[kichatna:13991] [28] /home/jed/petsc/ompi-optg/lib/libpetsc.so.3.015(+0xe86051)[0x7f75b79b2051]
[kichatna:13991] [29] /home/jed/petsc/ompi-optg/lib/libpetsc.so.3.015(SNESSolve+0x83a)[0x7f75b79e9e9a]
```

**Comment by jeremylt** [comment]
WRT caching, I was mostly concerned on the device side. I don't yet have a good sense of what can fill up a device.

**Comment by jeremylt** [comment]
I think I fixed it.I just updated CUDA, so I need to restart to check.

**Comment by jedbrown** [comment]
there's a memory leak now, and it rebuilds a kernel each iteration

**Comment by jeremylt** [comment]
Ok, I fixed the leaks I found. I think it should [finally] avoid recompilation.
```
./build/solids-elasticity -ceed /gpu/cuda -test -problem FSInitial-MR1 -mu_1 .5 -mu_2 .5 -nu 0.4 -degree 2 -dm_plex_box_faces 2,2,2 -num_steps 1 -bc_clamp 6 -bc_traction 5 -bc_traction_5 0,0,-.5 -expect_final_strain_energy 2.339138880207e-01 -snes_monitor
compiling...
compiling...
compiling...
compiling...
compiling...
compiling...
compiling...
compiling...
compiling...
compiling...
compiling...
compiling...
compiling...
compiling...
compiling...
compiling...
compiling...
compiling...
compiling...
compiling...
compiling...
compiling...
0 Load Increment
compiling...
  0 SNES Function norm 1.319444444444e-01 
compiling...
compiling...
compiling...
compiling...
compiling...
compiling...
compiling...
  1 SNES Function norm 1.269928720010e-01 
  2 SNES Function norm 1.619304232057e-01 
  3 SNES Function norm 2.596947950712e-01 
  4 SNES Function norm 1.056682943936e-01 
  5 SNES Function norm 3.874657827152e-02 
  6 SNES Function norm 1.691007474374e-03 
  7 SNES Function norm 3.777245889960e-05 
  8 SNES Function norm 3.049975250245e-10 
compiling...
```

Valgrind says now the only leaks are the usual CUDA messiness.

**Review by jedbrown — APPROVED** [review]
Looks good to me. Thanks.

**Pull request merged**